src/error.rs

```diff
@@ -7,123 +8,76 @@ use std::{
     io,
     path::PathBuf,
 };
-
-/// Wrapper around `serde_yaml::from_str`
+/// Shorthand for a Result that returns an `OrcaError`.
+pub type OrcaResult<T> = Result<T, OrcaError>;
+/// Possiable errors you may encounter.
 #[derive(Debug)]
-pub struct DeserializeFailure {
-    pub path: PathBuf,
-    pub error: serde_yaml::Error,
+pub enum OrcaError {
```

Comment:
In libraries like serce_yaml and std::io, it seems to be a common practice to have a rather simple `Error` struct that implement std::error and have the `enum` type be held internally. Do you know why they do this indirection?

Reply:
I also see this indirect Error struct to error Enum relationship found in library like tokio (https://docs.rs/tokio/latest/src/tokio/time/error.rs.html#26). This makes me believe that there is some decent reason to have this separation.

Reply:
Here is an explanation for why this separation is recommended:

----
This is a common pattern in Rust for error handling that offers several benefits. Let me explain the key advantages:

* Encapsulation and API Stability: 
  * By wrapping the Kind enum in an Error struct, the internal representation (Kind) can be changed without breaking public API compatibility
The pub(crate) on Kind means it's only visible within the crate, while Error is public
This allows the library to add new error kinds or restructure the internal representation without requiring users to update their code


* Controlled Error Creation:
  * The Error struct provides specific constructor methods (shutdown(), at_capacity(), invalid())
This gives more control over how errors are created and prevents users from directly creating invalid error states
The associated methods like is_shutdown() provide a clean API for error checking


* Pattern Matching Protection:
  * If Kind was directly exposed, users might pattern match on it, making their code brittle to changes
By hiding Kind, the library ensures users rely on the provided methods instead